File: r/sparklabs/batch/sales/sql/revenue_per_store.R
source("r/sparklabs/batch/sales/sql/utils.R")

showDF(run_sql("
  WITH revenue_per_order AS (
    SELECT oi.order_id,
           SUM(p.price * oi.quantity) AS total_revenue,
           COUNT(*) AS total_orders
    FROM order_items oi
    JOIN products p ON oi.product_id = p.product_id
    GROUP BY oi.order_id
  )
  SELECT o.store_id,
         SUM(r.total_revenue) AS store_revenue,
         COUNT(*) AS total_store_orders
  FROM orders o
  JOIN revenue_per_order r ON o.order_id = r.order_id
  JOIN stores s ON o.store_id = s.store_id
  GROUP BY o.store_id
  ORDER BY store_revenue DESC
"), truncate = FALSE)

File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/RevenuePerStore.scala
package io.devorbit.sparklabs.batch.sales

import io.devorbit.sparklabs.batch.sales.Utils
import org.apache.spark.sql.functions._

object RevenuePerStore extends App {

  private val revenuePerOrder = Utils.orderItems.join(Utils.products, Seq("product_id"), "inner")
    .withColumn("revenue", col("price") * col("quantity"))
    .groupBy(col("order_id"))
    .agg(sum("revenue").as("total_revenue"), count("*").as("total_orders"))
    .orderBy(desc("total_revenue"))

  Utils.orders
    .join(revenuePerOrder, Seq("order_id"), "inner")
    .join(Utils.stores, Seq("store_id"), "inner")
    .groupBy(col("store_id"))
    .agg(sum("total_revenue").as("store_revenue"), count("*").as("total_store_orders"))
    .orderBy(desc("store_revenue"))
    .show(false)

//  Thread.sleep(1000000)
}
